crypto/fipsmodule/ec/ec.c

```diff
@@ -506,17 +488,21 @@ int EC_GROUP_cmp(const EC_GROUP *a, const EC_GROUP *b, BN_CTX *ignored) {
     return 0;
   }
 
-  // |a| and |b| are both custom curves. We compare the entire curve
-  // structure. If |a| or |b| is incomplete (due to legacy OpenSSL mistakes,
-  // custom curve construction is sadly done in two parts) but otherwise not the
-  // same object, we consider them always unequal.
-  return a->meth != b->meth ||  //
-         !a->has_order || !b->has_order ||
-         BN_cmp(&a->order.N, &b->order.N) != 0 ||
-         BN_cmp(&a->field.N, &b->field.N) != 0 ||
-         !ec_felem_equal(a, &a->a, &b->a) ||  //
-         !ec_felem_equal(a, &a->b, &b->b) ||
-         !ec_GFp_simple_points_equal(a, &a->generator.raw, &b->generator.raw);
+  // |a| and |b| are both custom curves. We compare the entire curve structure
+  // if both |a| and |b| are complete. If either are incomplete (due to legacy
+  // OpenSSL mistakes, custom curve construction is sadly done in two parts
+  // |EC_GROUP_new_curve_GFp| -> |EC_GROUP_set_generator|), we only compare
+  // the parts that are available.
+  if (a->has_order && b->has_order) {
+    return a->meth != b->meth || BN_cmp(&a->order.N, &b->order.N) != 0 ||
+           BN_cmp(&a->field.N, &b->field.N) != 0 ||
+           !ec_felem_equal(a, &a->a, &b->a) ||
+           !ec_felem_equal(a, &a->b, &b->b) ||
+           !ec_GFp_simple_points_equal(a, &a->generator.raw, &b->generator.raw);
+  } else {
```

Comment:
in the `else` branch you can end up with one of the groups having a generator and the other one not. We should probably ask that either both groups have a generator (order) or both don't have it?

Reply:
This was intentional since I noticed that `ECTest.BIGNUMConvert` was failing. But after looking more closely, it's a test that I had wrote. The test is using an outdated generator to compare the point contents, I think we can just fix the test and add the additional check here.